Add Clown Corps comic source

altaccosc opened the pull request:
Checklist:

- [x] Updated `extVersionCode` value in `build.gradle` for individual extensions
- [x] Updated `overrideVersionCode` or `baseVersionCode` as needed for all multisrc extensions
- [x] Referenced all related issues in the PR body (e.g. "Closes #xyz")
- [x] Added the `isNsfw = true` flag in `build.gradle` when appropriate
- [x] Have not changed source names
- [x] Have explicitly kept the `id` if a source's name or language were changed
- [x] Have tested the modifications by compiling and running the extension through Android Studio
- [x] Have removed `web_hi_res_512.png` when adding a new extension

altaccosc committed:
Add ClownCorps comic source

AwkwardPeak7 commented on src/en/clowncorps/build.gradle:
```diff
@@ -0,0 +1,11 @@
+
+dependencies {
+    implementation(project(':lib:textinterceptor'))
+}
```
```suggestion
}

```
newline

AwkwardPeak7 commented on src/en/clowncorps/src/eu/kanade/tachiyomi/extension/en/clowncorps/ClownCorps.kt:
```diff
@@ -0,0 +1,158 @@
+    override val client: OkHttpClient
+        get() = super.client.newBuilder()
+            .addInterceptor(TextInterceptor())
+            .build()
```
```suggestion
    override val client = network.client.newBuilder()
            .addInterceptor(TextInterceptor())
            .build()
```

AwkwardPeak7 commented on src/en/clowncorps/src/eu/kanade/tachiyomi/extension/en/clowncorps/ClownCorps.kt:
```diff
@@ -0,0 +1,158 @@
+            //Image from: https://clowncorps.net/about/
+            thumbnail_url = "https://clowncorps.net/wp-content/uploads/2022/11/clowns41.jpg"
+            setUrlWithoutDomain("/comic")
+        }.let { Observable.just(MangasPage(listOf(it), false))!! }
```
```suggestion
    override fun fetchPopularManga(page: Int): Observable<MangasPage> = Observable.fromCallable {
        MangasPage(
            listOf(
                SManga.create().apply {
                    title = name
                    artist = creator
                    author = creator
                    description = synopsis
                    status = SManga.ONGOING
                    //Image from: https://clowncorps.net/about/
                    thumbnail_url = "https://clowncorps.net/wp-content/uploads/2022/11/clowns41.jpg"
                    setUrlWithoutDomain("/comic")
                }
            ),
            hasNextPage = false
        )
    }
```

AwkwardPeak7 commented on src/en/clowncorps/src/eu/kanade/tachiyomi/extension/en/clowncorps/ClownCorps.kt:
```diff
@@ -0,0 +1,158 @@
+        Observable.just(MangasPage(emptyList(), false))!!
+
+    override fun fetchMangaDetails(manga: SManga) =
+        Observable.just(manga.apply { initialized = true })!!
```
```suggestion
        Observable.just(manga.apply { initialized = true })
```

AwkwardPeak7 commented on src/en/clowncorps/src/eu/kanade/tachiyomi/extension/en/clowncorps/ClownCorps.kt:
```diff
@@ -0,0 +1,158 @@
+        }.let { Observable.just(MangasPage(listOf(it), false))!! }
+
+    override fun fetchSearchManga(page: Int, query: String, filters: FilterList) =
+        Observable.just(MangasPage(emptyList(), false))!!
```
```suggestion
    override fun fetchSearchManga(page: Int, query: String, filters: FilterList) = fetchPopularManga(page)
```

AwkwardPeak7 commented on src/en/clowncorps/src/eu/kanade/tachiyomi/extension/en/clowncorps/ClownCorps.kt:
```diff
@@ -0,0 +1,158 @@
+    override fun chapterListParse(response: Response): List<SChapter> {
+        val chapters = mutableListOf<SChapter>()
+
+        for (page in 1..Int.MAX_VALUE) {
```
get the last page from the response document

altaccosc replied:
Done in a1e1aca7b8a8de7127dd850fc3f2c51b4235d5fa

AwkwardPeak7 commented on src/en/clowncorps/src/eu/kanade/tachiyomi/extension/en/clowncorps/ClownCorps.kt:
```diff
@@ -0,0 +1,158 @@
+        chapters.forEachIndexed { index, sChapter ->
+            sChapter.chapter_number = chapters.size - index.toFloat()
+        }
+
```
app already populates the chapter number based on title, if it works fine then no need for this

altaccosc replied:
Done in f50c5421bed6f54c54cb13a40e19614efe9de4d8

AwkwardPeak7 commented on src/en/clowncorps/src/eu/kanade/tachiyomi/extension/en/clowncorps/ClownCorps.kt:
```diff
@@ -0,0 +1,158 @@
+            setOnPreferenceChangeListener { _, newValue ->
+                val checkValue = newValue as Boolean
+                preferences.edit().putBoolean(SHOW_AUTHORS_NOTES_KEY, checkValue).commit()
+            }
```
```suggestion
```
not needed, app already saves the preferences

altaccosc committed:
Apply suggestions from code review

Thank you very much!

Co-authored-by: AwkwardPeak7 <48650614+AwkwardPeak7@users.noreply.github.com>

altaccosc committed:
Only loop through the present pages

Don't just loop forever until a 404 is returned

altaccosc committed:
Disable reduntant sorting code

altaccosc committed:
Add date to chapters

AwkwardPeak7 commented on src/en/clowncorps/src/eu/kanade/tachiyomi/extension/en/clowncorps/ClownCorps.kt:
```diff
@@ -0,0 +1,183 @@
+        for (page in 1..pageCount) {
+            val url = "https://clowncorps.net/comic/page/$page/"
+            val resp = client.newCall(GET(url, headers)).execute()
+            if (resp.code == 404) break
```
```suggestion
```
remove

AwkwardPeak7 commented on src/en/clowncorps/src/eu/kanade/tachiyomi/extension/en/clowncorps/ClownCorps.kt:
```diff
@@ -0,0 +1,183 @@
+                    description = synopsis
+                    status = SManga.ONGOING
+                    // Image from: https://clowncorps.net/about/
+                    thumbnail_url = "https://clowncorps.net/wp-content/uploads/2022/11/clowns41.jpg"
```
```suggestion
                    thumbnail_url = "$baseUrl/wp-content/uploads/2022/11/clowns41.jpg"
```

AwkwardPeak7 commented on src/en/clowncorps/src/eu/kanade/tachiyomi/extension/en/clowncorps/ClownCorps.kt:
```diff
@@ -0,0 +1,183 @@
+        val chapters = mutableListOf<SChapter>()
+
+        for (page in 1..pageCount) {
+            val url = "https://clowncorps.net/comic/page/$page/"
```
```suggestion
            val url = "$baseUrl/comic/page/$page/"
```

AwkwardPeak7 commented on src/en/clowncorps/src/eu/kanade/tachiyomi/extension/en/clowncorps/ClownCorps.kt:
```diff
@@ -0,0 +1,183 @@
+        /*chapters.forEachIndexed { index, sChapter ->
+            sChapter.chapter_number = chapters.size - index.toFloat()
+        }*/
+
```
```suggestion
```
remove
also does it work without it?

altaccosc replied:
It works with it commented out, yes. I've been using it to read the comic for the past few days without any issues.
But I wanted to keep this snippet around in case the automatic sorting happens to break at some point.
It's not hurting anything by being there, is it?

stevenyomi replied:
This snippet forces chapter number to align with "sort by source order" in the app, so it's completely useless.

altaccosc replied:
With it commented out, the Sort By Source option sorts the chapters properly.
But the Sort By Chapter Number option sorts the pages reversed:
- chapter 1 page 42
- chapter 1 page 41
- (...)
- chapter 1 page 1
- chapter 1 cover
- chapter 2 page 73
- chapter 2 page 72
- etc...

altaccosc replied:
But sure, I can remove it completely, if you prefer.
I must admit I don't see much point in discussing a small commented out piece of code.

AwkwardPeak7 commented on src/en/clowncorps/src/eu/kanade/tachiyomi/extension/en/clowncorps/ClownCorps.kt:
```diff
@@ -0,0 +1,183 @@
+
+    private fun parseDate(dateStr: String): Long {
+        return try {
+            dateFormat.parse(dateStr)!!.time
```
```suggestion
            dateFormat.parse(dateStr)?.time ?: 0L
```

stevenyomi replied:
@AwkwardPeak7 No it won't be `null` at all, see [`Date parse(String) throws ParseException`](https://docs.oracle.com/javase/8/docs/api/java/text/DateFormat.html#parse-java.lang.String-). IDE wrongly prompts it as nullable because it wraps [`Date? parse(String, ParsePosition)`](https://docs.oracle.com/javase/8/docs/api/java/text/DateFormat.html#parse-java.lang.String-java.text.ParsePosition-) and throws on a check of `ParsePosition` so the IDE can't tell.

altaccosc committed:
Apply suggestions from code review

Co-authored-by: AwkwardPeak7 <48650614+AwkwardPeak7@users.noreply.github.com>

altaccosc committed:
Remove commented snippet

Co-authored-by: AwkwardPeak7 <48650614+AwkwardPeak7@users.noreply.github.com>

stevenyomi commented on src/en/clowncorps/src/eu/kanade/tachiyomi/extension/en/clowncorps/ClownCorps.kt:
```diff
@@ -0,0 +1,175 @@
+    override val name = "Clown Corps"
+    override val supportsLatest = false
+
+    private val creator = "Joe Chouinard"
```
Move to companion object.

altaccosc replied:
Done in 576a230f2a118b2bf12fae4fbf1256974af08105

stevenyomi commented on src/en/clowncorps/src/eu/kanade/tachiyomi/extension/en/clowncorps/ClownCorps.kt:
```diff
@@ -0,0 +1,175 @@
+
+    // Text from: https://clowncorps.net/about/
+    private val synopsis = "Clown Corps is a comic about crime-fighting clowns. " +
+        "It's pronounced \"core.\" Like marine corps."
```
Inline it.

altaccosc replied:
Done in 576a230f2a118b2bf12fae4fbf1256974af08105

stevenyomi commented on src/en/clowncorps/src/eu/kanade/tachiyomi/extension/en/clowncorps/ClownCorps.kt:
```diff
@@ -0,0 +1,175 @@
+        .addInterceptor(TextInterceptor())
+        .build()
+
+    override fun fetchPopularManga(page: Int): Observable<MangasPage> = Observable.fromCallable {
```
Use `Observable.just()`.

altaccosc replied:
Done in 0f555df0e9ed6801351247a6399cc608eed416b6

stevenyomi commented on src/en/clowncorps/src/eu/kanade/tachiyomi/extension/en/clowncorps/ClownCorps.kt:
```diff
@@ -0,0 +1,175 @@
+    override fun fetchPopularManga(page: Int): Observable<MangasPage> = Observable.fromCallable {
+        MangasPage(
+            listOf(
+                SManga.create().apply {
```
Extract to `getManga()` and use it in `fetchMangaDetails()`. Also add `initialized = true` in it.

altaccosc replied:
Done in 6dd186f4268be8df13a72c37617506f26a0ff005

stevenyomi commented on src/en/clowncorps/src/eu/kanade/tachiyomi/extension/en/clowncorps/ClownCorps.kt:
```diff
@@ -0,0 +1,175 @@
+
+    private val dateFormat by lazy {
+        SimpleDateFormat("MMMM dd, yyyy hh:mm aa", Locale.ENGLISH)
+    }
```
```suggestion
    override fun chapterListParse(response: Response): List<SChapter> {
        return response.asJsoup().select("#chapter-select .level-0").flatMap { chapterElement ->
            val url = chapterElement.attr("value")
            val document = client.newCall(GET(url, headers)).execute().asJsoup()
            document.select(".comic-list-dropdown-form .level-0").map {
                SChapter.create().apply {
                    setUrlWithoutDomain(it.attr("value"))
                    name = it.text()
                }
            }
        }.asReversed()
    }
```

altaccosc replied:
This does not set the date. Don't we want that?
All chapters are now all listed as having been posted today.

stevenyomi replied:
Your implementation is too slow - it will iterate through the 45 pages and send that many requests. That puts a lot of stress onto the source server.

altaccosc replied:
That's true and fair.
Is there nothing else that could be done for this, though?

altaccosc replied:
Oh, I just found something new!
Maybe we could loop over the years?
https://clowncorps.net/2024/?post_type=comic

stevenyomi replied:
Nothing much we can do. The current extension API doesn't provide the existing chapter list, so we can only fetch from scratch.

You can hack it by storing chapter info into shared preferences (considering you have only one manga entry) but code logic to do patch updates will be complicated and error-prone.

stevenyomi replied:
> Oh, I just found something new!\nMaybe we could loop over the years?\nhttps://clowncorps.net/2024/?post_type=comic

That's still 10 pages per webpage.

altaccosc replied:
Drat

altaccosc replied:
I added caching in 6ed7adf6be7170a20bdb913e9d115bbd18d67542, so hopefully that should lessen the load on the server enough to the point that my original method can still be used.

Quick summary of the way this caching works:
The first time, it loads all (45 as of the time of writing) webpages, and collects the chapters from those.
It then saves all of those chapters in the SharedPreferences, serialized to JSON.

The next time the comic gets refreshed, it gets all the chapters it can from the cache, and stores them in a **set**.
It then fetches every webpage starting at 1 again, which is where the most recent new chapters get added to.
Then, it continues going through every next page (thus going further and further into the past).
It keeps adding the newly fetched chapters into the set, but when all the chapters it just fetched are already in the set, it knows it has reached where the cache left off, so it stops the refresh.
If you regularly update, that means it only ever does one request.

Hopefully this explanation makes the code make a bit more sense.

I also added a button in the extension preferences menu to clear the cache.

altaccosc committed:
Move vars to companion obj & inline description

stevenyomi commented on src/en/clowncorps/src/eu/kanade/tachiyomi/extension/en/clowncorps/ClownCorps.kt:
```diff
@@ -26,14 +26,8 @@ import java.util.Locale
-    private val synopsis = "Clown Corps is a comic about crime-fighting clowns. " +
-        "It's pronounced \"core.\" Like marine corps."
+    override val name = NAME
+    override val supportsLatest = SUPPORTS_LATEST
```
Inline `name` and `supportsLatest`. I was asking you to move the creator constant only.

altaccosc replied:
Yeah I realised that too just now. Sorry, I don't use Kotlin a lot, so I'm not yet too caught-up on the exact conventions.
I'm learning quickly, though.
My next extensions will hopefully be better from the get-go.

altaccosc replied:
Done in 0f555df0e9ed6801351247a6399cc608eed416b6

altaccosc committed:
Un-move some constants & Use Observable.just

altaccosc committed:
Extract SManga creation to separate function

And use where necessary

stevenyomi commented on src/en/clowncorps/src/eu/kanade/tachiyomi/extension/en/clowncorps/ClownCorps.kt:
```diff
@@ -0,0 +1,168 @@
+        thumbnail_url = "$baseUrl/wp-content/uploads/2022/11/clowns41.jpg"
+        description = "Clown Corps is a comic about crime-fighting clowns.\n" +
+            "It's pronounced \"core.\" Like marine corps."
+        setUrlWithoutDomain("/comic")
```
```suggestion
        url = "/comic"
```

altaccosc committed:
Omit unnecessary function call

Co-authored-by: stevenyomi <95685115+stevenyomi@users.noreply.github.com>

altaccosc committed:
Add caching

I've tried a great many different ways of caching today, and as far as I can reason with my fried brain, I think this one now works pretty well.
I shall continue testing it on my phone.

stevenyomi commented on src/en/clowncorps/src/eu/kanade/tachiyomi/extension/en/clowncorps/ClownCorps.kt:
```diff
@@ -0,0 +1,241 @@
+
+            return name == other.name
+        }
+    }
```
```suggestion
    @Serializable
    class SerializableChapter(val fullLink: String, val name: String, val dateUpload: Long) {
        override fun hashCode() = fullLink.hashCode()
        override fun equals(other: Any?) = other is SerializableChapter && fullLink == other.fullLink
    }
```

Should use URL as identifier. And don't use `data class` if not needed.

stevenyomi commented on src/en/clowncorps/src/eu/kanade/tachiyomi/extension/en/clowncorps/ClownCorps.kt:
```diff
@@ -0,0 +1,241 @@
+        for (webpageIndex in 1..totalWebpageCount) {
+            val anyChaptersWereAdded = allChapters.addAll(fetchWebpageChapters(webpageIndex))
+            if (!anyChaptersWereAdded) break // No new chapters were added from this webpage, so we're done
+        }
```
Isn't the current response already the first page?

altaccosc replied:
Done in d2da1c0b5dc049f4bd998511b6941cd3e54f775a

stevenyomi commented on src/en/clowncorps/src/eu/kanade/tachiyomi/extension/en/clowncorps/ClownCorps.kt:
```diff
@@ -0,0 +1,241 @@
+        preferences.edit().putString(CACHE_KEY_CHAPTERS, fullJsonString).apply()
+
+        // Convert the serializable chapters to SChapters
+        return allChapters.map { chapter ->
```
How are you maintaining the chapters' order? It should be sorted from new to old. Either sort it by upload date (and hope post edits don't reset it) or maintain the order from the beginning.

altaccosc replied:
Done in ac05b17e338f4d0a5ebccce441b9b4fbba025135

stevenyomi commented on src/en/clowncorps/src/eu/kanade/tachiyomi/extension/en/clowncorps/ClownCorps.kt:
```diff
@@ -0,0 +1,241 @@
+    private fun getChaptersFromCache(): Set<SerializableChapter> {
+        val cachedChaps = preferences.getString(CACHE_KEY_CHAPTERS, null)
+        if (cachedChaps != null) return Json.decodeFromString(cachedChaps)
+        return emptySet()
```
```suggestion
        val cachedChaps = preferences.getString(CACHE_KEY_CHAPTERS, null) ?: return emptySet()
        return Json.decodeFromString(cachedChaps)
```

stevenyomi commented on src/en/clowncorps/src/eu/kanade/tachiyomi/extension/en/clowncorps/ClownCorps.kt:
```diff
@@ -0,0 +1,241 @@
+            val link = it.selectFirst(".post-title a")?.attr("href")
+            val title = it.selectFirst(".post-title a")?.text()
+            val postDate = it.selectFirst(".post-date")?.text()
+            val postTime = it.selectFirst(".post-time")?.text()
```
Any reason they could be missing? If not, use `!!`.

altaccosc replied:
Probably not, but in case they do happen to be missing, should it really crash at runtime instead of handling it properly?

stevenyomi replied:
It would cause the app to show an error instead, so you could identify the change on the website. Ignoring the error is not "handling it properly".

altaccosc replied:
Fair enough :+1:

altaccosc replied:
Done in 6cb092ddbb898d6163edb84230985b7ca40c5b6b

stevenyomi commented on src/en/clowncorps/src/eu/kanade/tachiyomi/extension/en/clowncorps/ClownCorps.kt:
```diff
@@ -0,0 +1,241 @@
+
+        // I couldn't find a way to create a simple button, so here's a workaround that uses
+        // a MultiSelectListPreference with a single option as a kind of confirmation window.
+        val clearCachePref = MultiSelectListPreference(screen.context).apply {
```
Using `SwitchPreferenceCompat` is enough (just don't confirm). And you don't need to specify a key for `SwitchPreferenceCompat` if you use it like a button.

altaccosc replied:
It works just as well, yes, but I feel like it's a less clear user experience.
This UX signifies that it's an option that toggles something, which it is not.
While using a pop-up, like I did previously, makes it feel completely seamless to the user.

I have tested this with some actual users, who all confirmed this.
I posted GIFs of both the toggle button (1), and the pop-up (2), which the question "What do you find clearer?"

> the second one because a toggle for something that isnt a toggle is weird

> By far the second option
> There is no world in which the first makes any sense
> I would consider option 1 broken software lol

The GIFs, for reference:
| Option 1 | Option 2|
|-----------|----------------|
| ![option1](https://github.com/keiyoushi/extensions-source/assets/97206249/fd9ec4f0-3839-49a4-b9a2-b00408be380a) | ![option2](https://github.com/keiyoushi/extensions-source/assets/97206249/aefb3a3f-4d25-4bb5-8cf6-3b2403eeb20e) |

stevenyomi replied:
Yeah but I'd prefer keeping the code as simple as possible. People don't visit the settings page very often, so it shouldn't need such a complicated workaround.

AwkwardPeak7 replied:
I would say having a good ux is important too

altaccosc replied:
I feel like we shouldn't compromise on UX, especially when the alternative option is as confusing as this.
And honestly, this code is not bad at all.
The only "odd" part is this:
```kotlin
setOnPreferenceChangeListener { _, newValue ->
    val checkValue = newValue as Set<*>
    if (checkValue.contains(VALUE_CONFIRM)) {
```
And this was adapted from how the MangaDex source does it: https://github.com/keiyoushi/extensions-source/blob/dfb59ebae0c8fa2b1a118e86b5bc5be8a183f3e6/src/all/mangadex/src/eu/kanade/tachiyomi/extension/all/mangadex/MangaDex.kt#L684-L685

stevenyomi replied:
Sure.

altaccosc committed:
Change SerializableChapter implementation

Co-authored-by: stevenyomi <95685115+stevenyomi@users.noreply.github.com>

altaccosc committed:
Code cleanliness

Co-authored-by: stevenyomi <95685115+stevenyomi@users.noreply.github.com>

altaccosc committed:
Optimise requests for page 1 away

altaccosc committed:
Explicitly sort chapters by upload date

altaccosc committed:
Move other preference options into separate functions

altaccosc committed:
Assume response document always contains what we're asking

And throw a runtime exception if it doesn't, so the problem can be noticed and fixed.
